Fix handling of '*' origin and add support of file:// protocol

pjanik opened the pull request:
Please look at the first commit only to see clean diff.

The goal of this PR is to handle pages served through file:// protocol. Behaviour varies in different browsers. Safari sets origin to 'file://' and everything works fine, but Chrome and Safari set message.origin to "null".
Also, https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage says:

> Lastly, posting a message to a page at a file: URL currently requires that the targetOrigin argument be "*".
> file:// cannot be used as a security restriction; this restriction may be modified in the future.

So, using `'*'` seems like the only possible solution to handle file:// protocol.
Also, iframe-phone didn't handle `'*'` at all, so I had to add it.
Finally, I did small cleanup and don't send `origin` in message body when it's not necessary (it's necessary only in hello msg sent from parent to iframe).

This change is backward compatible too.

scytacki commented on lib/iframe-endpoint.js:
```diff
@@ -82,7 +81,7 @@ function IFrameEndpoint() {
 
       // Perhaps-redundantly insist on checking origin as well as source window of message.
-      if (message.origin === parentOrigin) {
+      if (parentOrigin === '*' || message.origin === parentOrigin) {
```
I think it would be best to remove the parentOrigin check altogether. The iframe will currently communicate with any parentOrigin anyway. I know this goes against the recommendation of the MDN, but since our iframes will communicate with anything, this origin stuff adds some false sense of security. 

So the post call above would just `postToTarget(message, '*')`, and the parentOrigin variable could be removed entirely. 

For the time being, the parent-endpoint still needs to send `origin` in the hello message for backward compatibility.  But the file:// handling could be simplified.

scytacki commented on lib/parent-endpoint.js:
```diff
@@ -134,16 +133,32 @@ module.exports = function ParentEndpoint(targetWindowOrIframeEl, targetOrigin, a
+  // So, using '*' seems like the only possible solution.
+  if (selfOrigin === 'file://') {
+    selfOrigin = '*';
+  }
```
If the iframe-endpoint removes the use of parentOrigin. Then this handling of selfOrigin could be removed. Sometimes the parent will always send a origin as `file://`.  The targetOrigin handling would need to stay. The comment above should be expanded to explain how selfOrigin is ignored in version 1.2.0 and above.

pjanik commented:
Thanks! There's one more comment in my mailbox (removed?):

> The iframe will currently communicate with any parentOrigin anyway. I know this goes against the recommendation of the MDN, but since our iframes will communicate with anything, this origin stuff adds some false sense of security.

I didn't get it at first, but now I think I understand what you meant. You mean that it's impossible to provide explicit parentOrigin as an argument (but we'll always accept origin sent by parent), right? Also, "hot swap" of parent window (change of it's origin) seems impossible in any case. So, I agree that currently there's only false security and this check doesn't make sense. 

However another option would be to provide support of parentOrigin argument, which would default to '*'. And we'll never try to read it from parent's message. It'd be more correct approach if we think about a proper library that might be used by someone else too. We'll also reuse existing code, but we won't use this feature ourselves too soon probably.

Re parent endpoint - sending selfOrigin can be removed, agree. If targetOrigin is provided, we should probably use it. If it's not, our current mechanism only prevents from "hot swap" of iframe which seems very impossible to me. But theoretically it's doable, so well, we could leave it.

In general my impression is that origin check makes actually much more sense in iframe endpoint (where is not implemented) rather than parent endpoint. Example: https://community.hpe.com/t5/Protect-Your-Assets/XSS-and-App-Security-through-HTML5-s-PostMessage/ba-p/6515002
The question is if we care about those use cases or not.

pjanik commented:
@scytacki, so I've removed origin check from IframeEndpoint. I've added changelog to README and mentioned there file:// protocol.

It follows your recommendations and if we want extended security for IframeEndpoint, it can be easily implemented in v1.3.0.

scytacki commented:
Looks good.  

I agree, that adding an option for the user to provide an origin for IframeEndpoint would be good in the future. We might not use it, but it would make the security model more clear. For most of our stuff the issue in the XSS post you linked is not relevant because these are server-less interactives. There is no sensitive user data or cookies. The three exceptions are CODAP which has the document store in the backend, Model My Watershed, and the iSENSE interactive. So those ought to be audited. 

The comment you referred to is still there, it is just 'outdated' because it was referring to a line of code that has been changed. And yes you interpreted the comment correctly.
